codechecker_lib/database_handler.py

```diff
@@ -4,74 +4,379 @@
+        1, Derive from SQLServer and implement the abstract methods
+        2, Add/modify some command line options in CodeChecker.py
+        3, Modify SQLServer.from_cmdline_args() in order to create an
+           instance of the new server type if needed
+    '''
+
+    __metaclass__ = ABCMeta
+
+
+    def __init__(self, migration_root):
+        '''
+        Sets self.migration_root. migration_root should be the path to the
+        alembic migration scripts.
+        '''
+
+        self.migration_root = migration_root
+
+
+    def _create_or_update_schema(self, use_migration=True):
+        '''
+        Creates or updates the database schema. The database server should be
+        started before this method is called.
+
+        If use_migration is True, this method runs an alembic upgrade to HEAD.
+
+        In the False case, there is no migration support and only SQLAlchemy
+        meta data is used for schema creation.
+
+        On error sys.exit(1) is called.
+        '''
+
+        try:
+            db_uri = self.get_connection_string()
+            engine = sqlalchemy.create_engine(db_uri, strategy='threadlocal')
+
+            LOG.debug('Creating new database schema')
+            if use_migration:
+                LOG.debug('Creating new database session')
+                session = CreateSession(engine)
+                connection = session.connection()
+
+                cfg = config.Config()
+                cfg.set_main_option("script_location", self.migration_root)
+                cfg.attributes["connection"] = connection
+                command.upgrade(cfg, "head")
+
+                session.commit()
+            else:
+                CC_META.create_all(engine)
+
+            LOG.debug('Creating new database schema done')
+            return True
+
+        except sqlalchemy.exc.SQLAlchemyError as alch_err:
+            LOG.error(str(alch_err))
+            sys.exit(1)
+
+
+    @abstractmethod
+    def start(self, wait_for_start=True, init=False):
+        '''
+        Starts the database server and initializes the database server.
+
+        On wait_for_start == True, this method returns when the server is up
+        and ready for connections. Otherwise it only starts the server and
+        returns immediately.
+
+        On init == True, this it also initializes the database data and schema
+        if needed.
+
+        On error sys.exit(1) should be called.
+        '''
+        pass
+
+
+    @abstractmethod
+    def stop(self):
+        '''
+        Terminates the database server.
+
+        On error sys.exit(1) should be called.
+        '''
+        pass
+
+
+    @abstractmethod
+    def get_connection_string(self):
+        '''
+        Returns the connection string for SQLAlchemy.
+
+        DO NOT LOG THE CONNECTION STRING BECAUSE IT MAY CONTAIN THE PASSWORD
+        FOR THE DATABASE!
+        '''
+        pass
+
+
+    @classmethod
+    def from_cmdline_args(cls, args, context, env=None):
```

Comment:
arg:
should we use the whole arg here? 

context:
try not to depend too much on the context object, because it makes later testing/mocking harder.
right now both the postgresql and the sqlite uses only the codechecker_workspace, and migration_root values from it